bugtool/cmd/root.go

```diff
@@ -0,0 +1,297 @@
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/cilium/cilium/common"
+	"github.com/cilium/cilium/daemon/defaults"
+
+	"github.com/pierrre/archivefile/zip"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+// TODO: do we need filtering options?
+
+var RootCmd = &cobra.Command{
+	Use:   "bugtool",
+	Short: "retrieve information and files for debugging single Cilium node",
+	Long:  "TODO",
+	Run: func(cmd *cobra.Command, args []string) {
+		common.RequireRootPrivilege("bugtool")
+		runTool()
+	},
+}
+
+var (
+	serve    bool
+	port     int
+	dumpPath string
+)
+
+func init() {
+	RootCmd.Flags().BoolVar(&serve, "serve", false, "Serve captured files over HTTP")
+	RootCmd.Flags().IntVarP(&port, "port", "p",
+		4444, "Port to use when serving files, default is 4444")
+	RootCmd.Flags().StringVarP(&dumpPath, "tmp", "t", "/tmp/", "Path to store extraced files")
+}
+
+func runTool() {
+	var dbgDir string
+	// Prevent collision with other directories
+	d, err := ioutil.TempDir(dumpPath, "bugtool-")
+	if err != nil {
+		log.WithError(err).Warn("failed to create debug directory")
+	}
+	dbgDir = d
+
+	// TODO: review all error messages.
+	copyStateDir(dbgDir)
+	copySystemInfo(dbgDir)
+	// TODO: Copy relevent log files
+	// TODO: retrieve k8s log files
+
+	/*
+		Docker logs
+		access logs
+	*/
+
+	copyResolveConfig(dbgDir)
+	copyKernelConfig(dbgDir)
+	createArchive(dbgDir, dumpPath)
```

Comment:
How big is the file? I'm just trying to understand if it makes easier to paste everything into a single `.txt` file or multiple ones in a `.zip`. If it's a text file it's easier for users to just copy-paste it to the bug report on github.

Reply:
These files will be very big. We may want to have a summary which includes the essential information. We can have the bugtool concat that together based on the individual files.

Reply:
So the file is 110M on my workstation with no endpoints running, but this will be very host specific considering we are getting a lot of different command output. 

IMO the bugtool should gather as much information as it can. Then the user should pick what they are interested in. I don't think summarizing is a good idea. It's hard to know what exactly someone using the tool wants. If it gets too slow we can add filtering and / or use several archives as @aanm suggests.

Reply:
On my workstation the biggest files are `lsof` and `journalctl -u docker`. 
Also tried on a smaller docker deployment and there the archive is 5.2M and the biggest file is `cat /proc/kallsyms`.

Reply:
We probably don't need docker's logs. IMO `lsof` should be optional. We will probably get 95% of the information we want from `iptables-save`, `cilium logs`, `cilium endpoint list`, `uname -a`, `kubectl get version` and a couple more that I don't remember this moment.

Reply:
Dropping `lsof` then. Want to keep the docker logs just in case. Not sure yet of a good way to make these optional.

Reply:
@scanf why the docker logs?

Reply:
I was thinking it might capture useful information on environments where the cilium deployment is not called cilium but something else. Can drop it.